core/src/test/kotlin/io/github/kkafka/producer/TransactionKtTest.kt

```diff
@@ -0,0 +1,189 @@
+                mapOf(TopicPartition(topic, 0) to OffsetAndMetadata(31)),
+                "someGroupId"
+            )
+        }
+        verify(exactly = 1) { producer.commitTransaction() }
+        verify(exactly = 0) { producer.abortTransaction() }
+        assert(res == "someResult")
+    }
+
+    @Test
+    fun `transaction with single topic aborts and throws if consumer is not subscribed to given topic`() {
+        assertThrows<TransactionAbortedException> {
+            producer.transaction(consumer, "kf-some-unsubscribed-topic") {
+                // Block succeeds
+            }
+        }
+
+        verify(exactly = 1) { producer.beginTransaction() }
+        verify(exactly = 0) { producer.commitTransaction() }
+        verify(exactly = 1) { producer.abortTransaction() }
+    }
+
+    @Test
+    fun `transaction with single topic aborts and throws if block throws`() {
+        assertThrows<Exception> {
+            producer.transaction(consumer, topic) {
+                // Block fails
+                throw Exception()
+            }
+        }
+
+        verify(exactly = 1) { producer.beginTransaction() }
+        verify(exactly = 0) { producer.commitTransaction() }
+        verify(exactly = 1) { producer.abortTransaction() }
+    }
+
+    //
+    // Tests for transaction with no consumer
+    //
+
+    @Test
+    fun `no consumer transaction commits if block ran`() {
+        producer.transaction {
+            // Execution is fine
+        }
+
+        verify(exactly = 1) { producer.beginTransaction() }
+        verify(exactly = 1) { producer.commitTransaction() }
+        verify(exactly = 0) { producer.abortTransaction() }
+    }
+
+    @Test
+    fun `no consumer transaction aborts and throws if block threw`() {
+        assertThrows<Exception> {
+            producer.transaction {
+                // Execution fails
+                throw Exception()
+            }
+        }
+
+        verify(exactly = 1) { producer.beginTransaction() }
+        verify(exactly = 0) { producer.commitTransaction() }
+        verify(exactly = 1) { producer.abortTransaction() }
+    }
+
+    @Test
+    fun `no consumer transaction throws if commitTransaction() failed`() {
+        every { producer.commitTransaction() } throws (Exception())
+
+        assertThrows<Exception> {
+            producer.transaction {
+                // Execution succeeds
+            }
+        }
+
+        verify(exactly = 1) { producer.beginTransaction() }
+        verify(exactly = 1) { producer.commitTransaction() }
+        verify(exactly = 1) { producer.abortTransaction() }
+    }
+
+    @Test
+    fun `no consumer transaction returns block's return value`() {
+        val res = producer.transaction {
+            "hello"
+        }
+
+        verify(exactly = 1) { producer.beginTransaction() }
+        verify(exactly = 1) { producer.commitTransaction() }
+        verify(exactly = 0) { producer.abortTransaction() }
+
+        assert(res == "hello")
```

Comment:
Very minor, but in future PRs use our assertions library 
```kotlin
res shouldBe "hello"
```
for consistency and nicer messages when the test does fail. (It will report expected vs actual rather than just "assertion failed")